Cython/Utility/Coroutine.c

```diff
@@ -879,30 +956,58 @@ PyObject *__Pyx_PyGen_Send(PyGenObject *gen, PyObject *arg) {
 }
 #endif
 
-static CYTHON_INLINE
-PyObject *__Pyx_Coroutine_FinishDelegation(__pyx_CoroutineObject *gen) {
-    PyObject *ret;
+static CYTHON_INLINE __Pyx_PySendResult
+__Pyx_Coroutine_FinishDelegation(__pyx_CoroutineObject *gen, PyObject** retval) {
+    __Pyx_PySendResult result;
     PyObject *val = NULL;
     __Pyx_Coroutine_Undelegate(gen);
     __Pyx_PyGen__FetchStopIterationValue(__Pyx_PyThreadState_Current, &val);
     // val == NULL on failure => pass on exception
-    ret = __Pyx_Coroutine_SendEx(gen, val, 0);
+    result = __Pyx_Coroutine_SendEx(gen, val, retval, 0);
     Py_XDECREF(val);
-    return ret;
+    return result;
+}
+
+static __Pyx_PySendResult
+__Pyx_Coroutine_SendToDelegate(__pyx_CoroutineObject *gen, __Pyx_pyiter_sendfunc gen_am_send, PyObject *value, PyObject **retval) {
+    PyObject *ret;
+    __Pyx_PySendResult result;
+    // we assume that gen->yieldfrom cannot change as long as 'gen->is_running' is set => no safety INCREF()
+    gen->is_running = 1;
+    result = gen_am_send(gen->yieldfrom, value, &ret);
+    gen->is_running = 0;
+    if (result == PYGEN_NEXT) {
+        assert (ret != NULL);
+        *retval = ret;
+        return PYGEN_NEXT;
+    }
+    assert (result == PYGEN_ERROR ? ret == NULL : ret != NULL);
+    __Pyx_Coroutine_Undelegate(gen);
+    return __Pyx_Coroutine_SendEx(gen, ret, retval, 0);
 }
 
 static PyObject *__Pyx_Coroutine_Send(PyObject *self, PyObject *value) {
-    PyObject *retval;
-    __pyx_CoroutineObject *gen = (__pyx_CoroutineObject*) self;
-    PyObject *yf = gen->yieldfrom;
-    if (unlikely(gen->is_running))
-        return __Pyx_Coroutine_AlreadyRunningError(gen);
+    PyObject *retval = NULL;
+    __Pyx_PySendResult result = __Pyx_Coroutine_AmSend(self, value, &retval);
+    return __Pyx_Coroutine_MethodReturnFromResult(self, result, retval);
+}
 
-    if (yf) {
+static __Pyx_PySendResult
+__Pyx_Coroutine_AmSend(PyObject *self, PyObject *value, PyObject **retval) {
+    __Pyx_PySendResult result;
+    __pyx_CoroutineObject *gen = (__pyx_CoroutineObject*) self;
+    if (unlikely(gen->is_running)) {
+        *retval = __Pyx_Coroutine_AlreadyRunningError(gen);
+        return PYGEN_ERROR;
+    }
+    if (gen->yieldfrom_am_send) {
+        result = __Pyx_Coroutine_SendToDelegate(gen, gen->yieldfrom_am_send, value, retval);
+   } else if (gen->yieldfrom) {
+        PyObject *yf = gen->yieldfrom;
         PyObject *ret;
-        // FIXME: does this really need an INCREF() ?
-        //Py_INCREF(yf);
         gen->is_running = 1;
+        #if PY_VERSION_HEX < 0x030A00A3
```

Comment:
Probably `LIMITED_VERSION_HEX` I think, since we're not going to be able to retrieve `am_send` on limited API versions below that.

Reply:
Hmm. We don't _need_ these special cases if we use `am_send`. So, if we can't use `am_send` in Limited API < 3.10, then you're right, it would be better to have them. I'll change it.